### ISSUE TITLE
Imported libraries should be cached

### ISSUE THREAD

**Issue by pekkaklarck** [robotframework/robotframework#2106]
Currently we put imported libraries into an internal cache to avoid importing and processing them multiple times. It's questionable how much time that actually saves and there's a risk that we waste memory keeping references to libraries used only once. The cache also makes the overall import logic more complicated and causes bugs like #2105. Need to measure what kind of performance changes removing the cache would have.

We are also thinking about not caching resources (#2099) and variable files (#2101).

**Comment by yahman72** [comment]
This would be clearer indeed but could break things (triggered by #2165) - some things to consider :
1) use cache when imported with `Library` KW in `*** Settings ***`?
2) optionally use cache when imported with `BuiltIn.Import Library` KW (e.g. with `force_reimport=False` arg?

I have for example many resource files with possible duplicate imports Libs e.g.
a) `MyRes_A` uses `My_Lib_1` and `MyRes_C` 
b) `MyRes_B` uses `MyRes_C` 
c) `MyRes_C` uses `My_Lib_1`

Scenario a) would import `My_Lib_1` twice. This could become problematic if the `My_Lib_1` is stateful and is imported with `BuiltIn.Import Library`

**Comment by stasiek** [comment]
How about some flag-parameter for `Library` to make sure the import is fresh?
Like
`Library  SwingLibrary  FRESH`

**Comment by pekkaklarck** [comment]
I would rather have a solution that works out-of-the-box and doesn't require uses to know some special flag. Another problem is that libraries themselves can accept arguments and they can conflict with such special flags. We already have `WITH NAME` flag, but I would like to limit them only to cases where we really need them.

**Comment by pekkaklarck** [comment]
#2377 contains a related request to allow importing different libraries using same name. Need to think about these together.

**Comment by pekkaklarck** [comment]
#2931 is also somewhat related. It was closed because we already have this issue.

**Comment by pekkaklarck** [comment]
#3249 demonstrates another problem caused by library cache. Let's try to get this fixed in RF 3.2.

**Comment by pekkaklarck** [comment]
Should also look at #2377 when implementing this.

**Comment by pekkaklarck** [comment]
RF 3.2 is so long overdue that this is unfortunately descoped. If someone would really like this in, please comment here.

**Comment by pekkaklarck** [comment]
Removed from RF 5.0 scope simply because there are so many other more important enhancements. Another reason is that there probably aren't other changes that would affect the related code. Better to have a release where one of the main topics is enhancing importing and storing libraries as well as resource and variable files.

**Comment by franzhaas** [comment]
This is somewhat linked to this topic as well.: [#5122 ](https://github.com/robotframework/robotframework/issues/5122) however this is just a documentation clarification... much less impact
